<h1> Data Analytics Report </h1>

<h2> United Kingdom Road Accident Data Analysis </h2>
<h3> Inclusive Years: 2019 - 2022</h3>

<h1> Importing Libraries</h1>

In [1]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from scipy.stats import f_oneway

In [2]:
accident = pd.read_csv('dataset\\accident_data.csv')

In [3]:
accident

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious    05/06/2019  51.506187   
1       200701BS65737           Serious    02/07/2019  51.495029   
2       200701BS66127           Serious    26/08/2019  51.517715   
3       200701BS66128           Serious    16/08/2019  51.495478   
4       200701BS66837            Slight    03/09/2019  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    18/02/2022  57.374005   
660675  201091NM01881            Slight    21/02/2022  57.232273   
660676  201091NM01935            Slight    23/02/2022  57.585044   
660677  201091NM01964           Serious    23/02/2022  57.214898   
660678  201091NM02142           Serious    28/02/2022  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2                      NaN               Urban                    NaN   
3       Single carriageway               Urban     Fine no high winds   
4                      NaN               Urban                    NaN   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
4                              Other vehicle  
...                                      ...  
660674                                   Car  
660675                                   Car  
660676                                   Car  
660677                 Motorcycle over 500cc  
660

accident

<h1> Data Preparation</h1>

In [4]:
accident.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660679 entries, 0 to 660678
Data columns (total 14 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   Index                    660679 non-null  object 
 1   Accident_Severity        660679 non-null  object 
 2   Accident Date            660679 non-null  object 
 3   Latitude                 660654 non-null  float64
 4   Light_Conditions         660679 non-null  object 
 5   District Area            660679 non-null  object 
 6   Longitude                660653 non-null  float64
 7   Number_of_Casualties     660679 non-null  int64  
 8   Number_of_Vehicles       660679 non-null  int64  
 9   Road_Surface_Conditions  659953 non-null  object 
 10  Road_Type                656159 non-null  object 
 11  Urban_or_Rural_Area      660664 non-null  object 
 12  Weather_Conditions       646551 non-null  object 
 13  Vehicle_Type             660679 non-null  object 
dtypes: f

In [5]:
accident.describe()

Latitude      Longitude  Number_of_Casualties  Number_of_Vehicles
count  660654.000000  660653.000000         660679.000000       660679.000000
mean       52.553866      -1.431210              1.357040            1.831255
std         1.406922       1.383330              0.824847            0.715269
min        49.914430      -7.516225              1.000000            1.000000
25%        51.490690      -2.332291              1.000000            1.000000
50%        52.315641      -1.411667              1.000000            2.000000
75%        53.453452      -0.232869              1.000000            2.000000
max        60.757544       1.762010             68.000000           32.000000

In [6]:
accident.dtypes

Index                       object
Accident_Severity           object
Accident Date               object
Latitude                   float64
Light_Conditions            object
District Area               object
Longitude                  float64
Number_of_Casualties         int64
Number_of_Vehicles           int64
Road_Surface_Conditions     object
Road_Type                   object
Urban_or_Rural_Area         object
Weather_Conditions          object
Vehicle_Type                object
dtype: object

<h1> Data Cleansing </h1>

In [7]:
accident.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                      25
Light_Conditions               0
District Area                  0
Longitude                     26
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions      726
Road_Type                   4520
Urban_or_Rural_Area           15
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

In [8]:
accident['Latitude'] = accident['Latitude'].fillna(accident['Latitude'].mode()[0])
accident['Longitude'] = accident['Longitude'].fillna(accident['Longitude'].mode()[0])
accident['Road_Surface_Conditions'] = accident['Road_Surface_Conditions'].fillna('Road_Surface_Conditions')
accident['Road_Type'] = accident['Road_Type'].fillna('Road_Type')
accident['Weather_Conditions'] = accident['Weather_Conditions'].fillna('Weather_Conditions')
accident['Urban_or_Rural_Area'] = accident['Urban_or_Rural_Area'].fillna(accident['Urban_or_Rural_Area'].mode()[0])

In [9]:
accident.isnull().sum()

Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
dtype: int64

<h1> Converting Object to DataTime Data Time </h1>

In [10]:
accident['Accident Date'] = pd.to_datetime(accident['Accident Date'], dayfirst = True)

In [11]:
accident.dtypes

Index                              object
Accident_Severity                  object
Accident Date              datetime64[ns]
Latitude                          float64
Light_Conditions                   object
District Area                      object
Longitude                         float64
Number_of_Casualties                int64
Number_of_Vehicles                  int64
Road_Surface_Conditions            object
Road_Type                          object
Urban_or_Rural_Area                object
Weather_Conditions                 object
Vehicle_Type                       object
dtype: object

<h1> Extracting date information using pandas date time </h1>

In [12]:
accident['Year'] = accident['Accident Date'].dt.year
accident['Month'] = accident['Accident Date'].dt.month
accident['Day'] = accident['Accident Date'].dt.day
accident['DayOfWeek'] = accident['Accident Date'].dt.dayofweek

In [13]:
accident.isnull().sum()

Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
Year                       0
Month                      0
Day                        0
DayOfWeek                  0
dtype: int64

In [14]:
#accident['Accident Date'] = accident['Accident Date'].str.strip()
#accident['AccidentDate'] = accident['AccidentDate'].str.replace('/', '-')

<h1>Data Analytics</h1>

<h1>Analyzing Each from the Data Set</h1>

<h2>Univariate</h2>

In [15]:
df_2019 = accident[accident['Year'] == 2019]
df_2019

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious    2019-06-05  51.506187   
1       200701BS65737           Serious    2019-07-02  51.495029   
2       200701BS66127           Serious    2019-08-26  51.517715   
3       200701BS66128           Serious    2019-08-16  51.495478   
4       200701BS66837            Slight    2019-09-03  51.488576   
...               ...               ...           ...        ...   
182110       2.01E+12            Slight    2019-12-20  54.985289   
182111       2.01E+12           Serious    2019-12-21  54.984105   
182112       2.01E+12            Slight    2019-12-23  55.166369   
182113       2.01E+12            Slight    2019-12-23  54.995154   
182114       2.01E+12           Serious    2019-12-22  55.104618   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
182110       Darkness - no lighting   Dumfries and Galloway  -3.210294   
182111                     Daylight   Dumfries and Galloway  -3.193693   
182112       Darkness - no lighting   Dumfries and Galloway  -2.992068   
182113        Darkness - lights lit   Dumfries and Galloway  -3.058338   
182114                     Daylight   Dumfries and Galloway  -3.348857   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
182110                     1                   1            Frost or ice   
182111                     2                   1            Frost or ice   
182112                     1                   1            Frost or ice   
182113                     1                   1             Wet or damp   
182114                     2                   2            Frost or ice   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2                Road_Type               Urban     Weather_Conditions   
3       Single carriageway               Urban     Fine no high winds   
4                Road_Type               Urban     Weather_Conditions   
...                    ...                 ...                    ...   
182110  Single carriageway               Rural                  Other   
182111  Single carriageway               Rural                  Other   
182112  Single carriageway               Rural            Fog or mist   
182113  Single carriageway               Rural     Fine no high winds   
182114  Single carriageway               Rural  Raining no high winds   

                                Vehicle_Type  Year  Month  Day  DayOfWeek  
0                                        Car  2019      6    5          2  
1                                        Car  2019      7    2          1  
2                      Taxi/Private hire car  2019      8   26          0  
3       Bus or coach (17 or more pass seats)  2019      8   16          4  
4                              Other vehicle  2019      9    3          1  
...                                      ...   ...    ...  ...  

<h3>Q: How many incidents happened under light condition</h3>

In [16]:
df_2019 ['Light_Conditions'].value_counts()

Light_Conditions
Daylight                       133042
Darkness - lights lit           35769
Darkness - no lighting          10842
Darkness - lighting unknown      1768
Darkness - lights unlit           694
Name: count, dtype: int64

<h1>Insight 1: How many incidents happened in Highland? </h1>

In [42]:
highland_incidents = accident[accident['District Area'] == 'Highland'].shape[0]
print(f'{highland_incidents} incidents happened in Highland.')

2021 incidents happened in Highland.


<h1>Insights 2: There is 36378 incidents happened in Highlind, Now, What is the average number of casualties happened in Highland on 2020?</h1>

In [48]:
accident_2020 = accident[(accident['District Area'] == 'Highland') & (accident['Year'] == 2020)]
mean_casualties = round(accident_2020['Number_of_Casualties'].mean(),2)
print(f"The average number of casualties in Highland in 2020 was {mean_casualties}.")

The average number of casualties in Highland in 2020 was 1.44.


<h1> Insight 3: The average number of casualties happened in Highland on 2020 is 1.44, What is the Average Number of casualties happened in Kensington and Chelsea on 2019? </h1>

In [49]:
accident_2019 = accident[(accident['District Area'] == 'Kensington and Chelsea') & (accident['Year'] == 2019)]
mean_casualties = round(accident_2019['Number_of_Casualties'].mean(),2)
print(f"The average number of casualties in Kensington and Chelsea in 2019 was {mean_casualties}.")

The average number of casualties in Kensington and Chelsea in 2019 was 1.15.


<h1> Insight 4: What is the average Number of casualties happened in Dumfries and Galloway on 2021?</h1>

In [51]:
accident_2021 = accident[(accident['District Area'] == 'Dumfries and Galloway') & (accident['Year'] == 2021)]
mean_casualties = round(accident_2021['Number_of_Casualties'].mean(),2)
print(f"The average number of casualties in Dumfries and Galloway in 2021 was {mean_casualties}.")

The average number of casualties in Dumfries and Galloway in 2021 was 1.37.


<h1> Insights 5: Which vehicle type is most involved in accidents? 

In [56]:
accident_Road = accident.groupby(['Vehicle_Type'])['Number_of_Casualties'].count()
print(f"The Car has the most number of casualties happened during Accident.{accident_Road}")


The Car has the most number of casualties happened during Accident.Vehicle_Type
Agricultural vehicle                       1947
Bus or coach (17 or more pass seats)      25878
Car                                      497992
Data missing or out of range                  6
Goods 7.5 tonnes mgw and over             17307
Goods over 3.5t. and under 7.5t            6096
Minibus (8 - 16 passenger seats)           1976
Motorcycle 125cc and under                15269
Motorcycle 50cc and under                  7603
Motorcycle over 125cc and up to 500cc      7656
Motorcycle over 500cc                     25657
Other vehicle                              5637
Pedal cycle                                 197
Ridden horse                                  4
Taxi/Private hire car                     13294
Van / Goods 3.5 tonnes mgw or under       34160
Name: Number_of_Casualties, dtype: int64


<h1> Insight 6: Do larger vehicle type influence the accidents in which road type?

In [59]:
accident_Road = accident.groupby(['Vehicle_Type','Road_Type'])['Number_of_Casualties'].count()
print("Bigger vehicles might be more likely to have accidents on certain roads.")
accident_Road

Bigger vehicles might be more likely to have accidents on certain roads.


Vehicle_Type                         Road_Type         
Agricultural vehicle                 Dual carriageway        293
                                     One way street           49
                                     Road_Type                13
                                     Roundabout              129
                                     Single carriageway     1436
                                                           ...  
Van / Goods 3.5 tonnes mgw or under  One way street          723
                                     Road_Type               215
                                     Roundabout             2294
                                     Single carriageway    25460
                                     Slip road               361
Name: Number_of_Casualties, Length: 87, dtype: int64

<h1>Insights 7: The average latitude is  52.55, suggesting that the data is concentrated in a specific region. The Standard deviation of 1.40 suggest a wide spread of latitue. </h1>

In [23]:
avg_Latitude = accident['Latitude'].mean()
std_Latitude = accident['Latitude'].std()
avg_Latitude
std_Latitude

np.float64(1.4068956719685075)

<h1>Insight 8: The average longitude is -1.43, indicating a location in the western hemisphere. The standard deviation of 1.38 is similar spread to latitude</h1>

In [24]:
avg_Longitude = accident['Longitude'].mean()
std_Longitude = accident['Longitude'].std()
std_Longitude

np.float64(1.3833052405598685)

<h1>Insights 9: The average number of casualties is 1.36, indicating that most accidents involve a small number of casualties. The Standard deviation of 0.82 refers that a significant variation in the number of casualties.</h1>

In [25]:
avg_casualty = accident['Number_of_Casualties'].mean()
std_casualty = accident['Number_of_Casualties'].std()
avg_casualty
std_casualty

np.float64(0.8248469447164726)

<h1>Insight 10: What is the maximum number of casualties reported in each year?</h1>

In [26]:
max_casualty = accident.groupby(accident['Year'])['Number_of_Casualties'].max()
max_casualty
print(f'The maximum number of casualties happended each year are {max_casualty}:')

The maximum number of casualties happended each year are Year
2019    68
2020    62
2021    48
2022    43
Name: Number_of_Casualties, dtype: int64:


<h1>Insight 11: What is the relationship between the number of casualties and vehicles involved over time?</h1>

In [27]:
monthly_casualty = accident.groupby(['Year','Month'])[['Number_of_Casualties', 'Number_of_Vehicles']].sum().unstack()
monthly_casualty
print(f'The number of casualties lessen from year 2019 to 2022.{monthly_casualty}')

The number of casualties lessen from year 2019 to 2022.      Number_of_Casualties                                                   \
Month                   1      2      3      4      5      6      7      8    
Year                                                                          
2019                 20762  17872  20349  19232  21747  21209  21846  21007   
2020                 18940  19189  18337  18286  19422  19129  19761  18680   
2021                 18173  14648  17815  17335  18852  18728  19682  18797   
2022                 13163  14804  16575  15767  16775  17230  17201  16796   

                     ... Number_of_Vehicles                              \
Month     9      10  ...                 3      4      5      6      7    
Year                 ...                                                  
2019   20709  20877  ...              27707  26107  29213  28714  29372   
2020   19388  21029  ...              24532  24817  26145  26065  26998   
2021   18456  20

<h1>Insights 12: What is the average number of casualties per month?</h1>

In [60]:
avg_casualty_month = accident.groupby(accident['Month'])['Number_of_Casualties'].mean().sum()
print(f'The evarage number of casualties per month is:{avg_casualty_month}')

The evarage number of casualties per month is:16.28566100292791


<h1>Insights 13: Does number of casualties has changed over time?</h1>

In [64]:
year_casualty = accident.groupby(accident['Year'])['Number_of_Casualties'].sum()
print('The number of accidents has been decreasing each year.')
year_casualty

The number of accidents has been decreasing each year.


Year
2019    247780
2020    230905
2021    222146
2022    195737
Name: Number_of_Casualties, dtype: int64

<h1>Insight 14: Are there other variables that might be more strongly correlated with the number of casualties, such as vehicle type, road conditions, or driver behavior?</h1>

In [67]:
casualty_vehicle = accident['Number_of_Casualties'].corr(accident['Number_of_Vehicles'])
rounded_correlation = round(casualty_vehicle, 2)
print(f'{rounded_correlation} There is correlation between the number of casualties and the number of vehicles but it is weak.')

0.23 There is correlation between the number of casualties and the number of vehicles but it is weak.


<h1>Insight 15:Why are accidents more severe during low-light conditions?</h1>

In [70]:
severity_light = accident.groupby('Light_Conditions')['Accident_Severity'].value_counts().unstack()
print('Accidents are more severe during low-light conditions.')
severity_light

Accidents are more severe during low-light conditions.


Accident_Severity            Fatal  Serious  Slight
Light_Conditions                                   
Darkness - lighting unknown     68      794    5622
Darkness - lights lit         1860    19130  108345
Darkness - lights unlit         45      360    2138
Darkness - no lighting        1612     7174   28651
Daylight                      5076    60759  419045

<h1>Insights 16: How accident severity varies with different weather conditions?</h1>

In [71]:
severity_weather = accident.groupby('Weather_Conditions')['Accident_Severity'].value_counts().unstack()
print('While accidents are more frequent in rainy weather, accidents occurring in snowy conditions are significantly more likely to result in serious injuries or fatalities.')
severity_weather

While accidents are more frequent in rainy weather, accidents occurring in snowy conditions are significantly more likely to result in serious injuries or fatalities.


Accident_Severity      Fatal  Serious  Slight
Weather_Conditions                           
Fine + high winds        175     1245    7134
Fine no high winds      7100    72046  441739
Fog or mist               82      483    2963
Other                    165     1801   15184
Raining + high winds     145     1261    8209
Raining no high winds    848     9468   69380
Snowing + high winds       3      109     773
Snowing no high winds     36      565    5637
Weather_Conditions       107     1239   12782

<h1>Insights 17: What are the District and road types with the highest accidents frequencies?</h1>

In [76]:
accident_counts = accident.groupby(['District Area', 'Road_Type'])['Index'].count().unstack()
highest_district = accident_counts.sum(axis=1).idxmax()
highest_road_type = accident_counts.loc[highest_district].idxmax()
print(f'While {highest_district} has the highest overall number of accidents, {highest_district} experiences a disproportionately high number of accidents on {highest_road_type}, suggesting a potential need for targeted safety improvements in that specific area.')

While Birmingham has the highest overall number of accidents, Birmingham experiences a disproportionately high number of accidents on Single carriageway, suggesting a potential need for targeted safety improvements in that specific area.


<h1>Insights 18: Areas with higher casualty rates per accident, suggesting potentially more dangerous road conditions. </h1>

In [77]:
avg_casualties = accident.groupby(['District Area', 'Road_Type'])['Number_of_Casualties'].mean()
highest_casualty_area = avg_casualties.idxmax()
highest_casualty_rate = avg_casualties.max()
print(f"The district and road type with the highest average number of casualties per accident is {highest_casualty_area} with an average of {highest_casualty_rate} casualties per accident. This suggests potentially more dangerous road conditions in that specific area.")

The district and road type with the highest average number of casualties per accident is ('Corby', 'One way street') with an average of 4.0 casualties per accident. This suggests potentially more dangerous road conditions in that specific area.


<h1>Insights 19: Which combinations of weather and road surface conditions lead to the highest average number of casualties?</h1>

In [78]:
avg_casualties = accident.groupby(['Weather_Conditions', 'Road_Surface_Conditions'])['Number_of_Casualties'].mean()
highest_casualty_combo = avg_casualties.idxmax()
highest_casualty_rate = avg_casualties.max()
print(f"The combination of weather and road surface conditions with the highest average number of casualties per accident is {highest_casualty_combo} with an average of {highest_casualty_rate} casualties per accident. This suggests that driving under these conditions poses a significant risk.")

The combination of weather and road surface conditions with the highest average number of casualties per accident is ('Snowing + high winds', 'Dry') with an average of 2.5 casualties per accident. This suggests that driving under these conditions poses a significant risk.


<h1> Insights 20: Districts with high accident rates in either Urban or Rural Areas.

In [36]:
district_counts = accident.groupby(['District Area', 'Urban_or_Rural_Area'])['Index'].count().unstack()
district_counts

Urban_or_Rural_Area   Rural  Unallocated   Urban
District Area                                   
Aberdeen City         207.0          NaN  1116.0
Aberdeenshire        1781.0          NaN   149.0
Adur                  197.0          NaN   422.0
Allerdale             880.0          NaN   248.0
Alnwick               232.0          NaN     NaN
...                     ...          ...     ...
Wychavon             1088.0          NaN   273.0
Wycombe               739.0          NaN   990.0
Wyre                  480.0          NaN   758.0
Wyre Forest           419.0          NaN   550.0
York                  706.0          NaN  1191.0

[422 rows x 3 columns]